rust/datafusion/src/physical_plan/expressions.rs

```diff
@@ -1985,9 +1982,10 @@ mod tests {
 
     #[test]
     fn invalid_cast() -> Result<()> {
-        let schema = Schema::new(vec![Field::new("a", DataType::Utf8, false)]);
-        let result = cast(col("a"), &schema, DataType::Int32);
-        result.expect_err("Invalid CAST from Utf8 to Int32");
+        // Ensure a useful error happens at plan time if invalid casts are used
```

Comment:
It turns out that arrow can, in fact, cast from utf8 -> Int32

Reply:
yup, unifying the cast logic was good for this. I've wanted to add cast options, such as disallowing lossy casts.
If/when we get to that point, we'll have to think about what behaviour we want DataFusion to use.

Reply:
> If/when we get to that point, we'll have to think about what behaviour we want DataFusion to use.

I think DataFusion now makes the distinction between "casting" (aka if the user specifically requests to cast from one type to another) which can be lossy and "coercion" (aka when casts need to be added explicitly so that expressions can be evaluated (e.g. plus). 

Coercion is designed not be lossy, but casting can be. 

The downside is that Datafusion has to have another set of rules of what type coercions are allowed (e.g. I need to add https://github.com/apache/arrow/pull/8463 to properly support DictionaryArray in DataFusion).